crates/circuit/src/circuit_instruction.rs

```diff
@@ -63,15 +63,37 @@ pub struct CircuitInstruction {
     pub clbits: Py<PyTuple>,
 }
 
+impl CircuitInstruction {
+    pub fn new<T1, T2, U1, U2>(
+        py: Python,
+        operation: PyObject,
+        qubits: impl IntoIterator<Item = T1, IntoIter = U1>,
+        clbits: impl IntoIterator<Item = T2, IntoIter = U2>,
+    ) -> PyResult<Py<Self>>
```

Comment:
If this is the rust space constructor why are we putting it in a `Py<>`?

Reply:
It's an interesting point. It'd probably be a better factoring to just return `Self`, but the only callers right now stuff it into a `Py` so I did it here for less code.

I think there's a much larger conversation we should have around the separation we'd ultimately want between the Python and Rust (model) hierarchies. Especially with the foresight that we'll probably want a pure Rust API at some point, which would likely mean having core Rust-only structs and then PyO3 wrappers around those.

Personally, I'd like the `qiskit-circuit` crate not depend on PyO3 at all, and then have a new crate `qiskit-pyo3` that depends on PyO3 and `qiskit-circuit` (and other core Rust-only crates) to bridge everything together. Before we get to that point, I'm not sure it matters too much for `new` versus `py_new` to be entirely Python versus entirely Rust, since I'd think we'd want to end up with two separate structs for each case, anyway.

Reply:
Done in 84fad3e252395154a3d47d3555cfdab84e6f4293.